fasm2bels/lib/interchange_capnp.py

```diff
@@ -0,0 +1,789 @@
+        net_name = flatten_wires(source_wire, top.wire_assigns,
+                                 top.wire_name_net_map)
+        if sink_wire in top.wire_name_net_map:
+            assert top.wire_name_net_map[sink_wire] == net_name
+        else:
+            top.wire_name_net_map[sink_wire] = net_name
+
+    # Create a list of each primative instances to later build up a primative
+    # model library.
+    hdi_primitives_cells = {}
+
+    # Create cells instances from each bel in the design.
+    for site in top.sites:
+        for bel in sorted(site.bels, key=lambda bel: bel.priority):
+            bel.output_interchange(
+                top_cell=top_cell,
+                top=top,
+                net_map=top.wire_name_net_map,
+                constant_nets=constant_nets,
+            )
+
+            if bel.module not in hdi_primitives_cells:
+                hdi_primitives_cells[bel.module] = []
+
+            hdi_primitives_cells[bel.module].append(bel)
+
+    # Add top level cell to the work cell library.
+    work.add_cell(top_cell)
+
+    # Construct library cells based on data from top module.
+    for cellname in hdi_primitives_cells:
+        instances = hdi_primitives_cells[cellname]
+
+        cell = Cell(cellname)
+
+        ports = {}
+        for instance in instances:
+            _, connections, port_is_output = instance.create_connections(top)
+
+            for port in connections:
+                if port_is_output[port]:
+                    # The current model doesn't handle IO at all, so add
+                    # special cases for IO ports in the library.
+                    if cellname.startswith('IOBUF') and port == "IO":
+                        direction = Direction.Inout
+                    else:
+                        direction = Direction.Output
+                else:
+                    direction = Direction.Input
+
+                width = connections[port].bus_width()
+                if port in instance.port_width:
+                    if width is not None:
+                        assert width <= instance.port_width[port], port
+
+                    width = instance.port_width[port]
+
+                if port in ports:
+                    port_dir, port_width = ports[port]
+                    assert port_dir == direction
+                    if width is not None:
+                        assert port_width <= width
+
+                        if width > port_width:
+                            ports[port] = (direction, width)
+                    else:
+                        assert port_width is None
+                else:
+                    ports[port] = (direction, width)
+
+        for port, (direction, width) in ports.items():
+
+            if width is not None:
+                cell.add_bus_port(port, direction, start=0, end=width - 1)
+            else:
+                cell.add_port(port, direction)
+
+        hdi_primitives.add_cell(cell)
```

Comment:
> I presume that the library has to contain cell definitions not instances

Correct

> If that is the case you will overwrite here existing cells with new ones without checking whether their definitions are identical.

Why?  `hdi_primitives` was empty before line 614, and `hdi_primitives_cells` should contain the unique list of primatives in the logical netlist.

Reply:
I mean that here you loop over all cell instances and use them to make the library. What if you encounter two cells with the same type name and eg. different port names/widths? I know that it should never happen in a correct netlist but may happen as a result of a bug of some sort.

I imagine that the logic of adding a cell to the library would be: if a cell definition with the given type name does not exist then add it but if it does then check whether it matches the new one added. Raise an error if it doesn't.

Reply:
Take another look at the loop, it isn't quiet as you say.  hdi_primitives_cells is a dict of primitive cell types (e.g. LUT6_2, RAM64X1D), so the outer loop is simply per primitive.  The inside loop is then over each instance of the primitive, and the library definition it builds is the widest of those.  There is also an explicit `set_port_width`, which is needed when there is ambiguity around the width of the primitive (e.g. RAMB18)

It is also worth noting at some point we should just use the data from the Unisim library for this, rather than inference.

Reply:
Ok, I get it now. Thanks.